microcontroller/Core/Src/frequency.c

```diff
@@ -0,0 +1,59 @@
+#include "stm32f4xx_hal.h"
+#include "stdio.h"
+#include "math.h"
+
+// Sensor input pin
+#define SENSOR_PIN GPIO_PIN_0
+#define SENSOR_PORT GPIOA
+
+//Threshold values
+#define OUTLIER_COUNT 2
+#define OUTLIER_THRESHOLD 0.1
+
+uint32_t current_counter = 0;
+uint32_t last_counter = 0;
+uint32_t Difference = 0;
+uint32_t outlier_counter = 0;
+uint32_t freq = 0;
+uint32_t current_freq = 0;
+
+
+bool check_outliers(uint32_t curr_freq, uint32_t incoming_freq) {
+	uint32_t difference = incoming_freq > curr_freq ? incoming_freq - curr_freq: curr_freq - incoming_freq;
+	float average = (incoming_freq + curr_freq) / 2.0;
+	float percent_difference = (difference / average) * 100.0;
+	if (percent_difference > OUTLIER_THRESHOLD) {
+	    return true;
+	} else {
+		return false;
+	}
+}
+
+void read_freq() {
```

Comment:
For the read frequency function I have an implemented function from my previous microcontroller code that I can give you. I'd probably just recommend using mine just because it's been thoroughly tested. I'll give it to you during workshop.